[docs] Automatic user provisioning for MySQL (#33084)
gravitational/teleport: docs/pages/database-access/auto-user-provisioning/mysql.mdx

---
title: MySQL Automatic User Provisioning 
description: Configure automatic user provisioning for MySQL.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="14.1"
>
  Automatic user provisioning for MySQL is available starting from Teleport
  14.1.
</Details>

## Prerequisites

- Teleport cluster with a configured [self-hosted MySQL](../guides/mysql-self-hosted.mdx)
  or [RDS MySQL](../guides/rds.mdx) database.
- Ability to connect to and create user accounts in the target database.

<Admonition type="note" title="Supported versions">

> Automatic user provisioning are not compatible with MySQL versions lower than

Reviewer: ```suggestion
Automatic user provisioning is not compatible with MySQL versions lower than
```

8.0.
</Admonition>

## Step 1/3. Configure database admin

(!docs/pages/includes/database-access/auto-user-provisioning-configure-admin.mdx!)

> Teleport use the same authentication mechanism when connecting as an admin user

Reviewer: ```suggestion
Teleport uses the same authentication mechanism when connecting as an admin user
```

as for regular user connections: X.509 for self-hosted databases and AWS IAM
for RDS.

The admin user must have privileges within the database to create users and
grant them privileges. The admin user must also have privileges to monitor user
processes and role assignments.

> In addition, a database/schema is required for the admin user to logs into by

Reviewer: ```suggestion
In addition, a database/schema is required for the admin user to log into by
```

> In addition, a database/schema is required for the admin user to logs into by

Reviewer: What does "database/schema" mean here?

Author: Ya. It's a database inside the MySQL server, like the <db-name> in `CREATE DATABASE <db-name>` or `tsh db connect --db-name <db-name>`.  Schema and database are synonym for MySQL. 

I thought it would be unclear to say a `database is required xxx`. Maybe I should just leave it as `schema`?

Reviewer: I think `a schema is required` works!

authentication:
```sql
CREATE USER 'teleport-admin' IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS';
GRANT SELECT ON mysql.role_edges TO 'teleport-admin' ;
GRANT PROCESS, ROLE_ADMIN, CREATE USER ON *.* TO 'teleport-admin' ;

CREATE DATABASE IF NOT EXISTS `teleport`;
GRANT ALTER ROUTINE, CREATE ROUTINE, EXECUTE ON `teleport`.* TO 'teleport-admin' ;
```
</TabItem>

<TabItem label="Self-hosted MySQL">
The self-hosted MySQL admin user must have X.509 authentication configured:
```sql
CREATE USER "teleport-admin" REQUIRE SUBJECT "/CN=teleport-admin";
GRANT SELECT ON mysql.role_edges TO 'teleport-admin' ;
GRANT PROCESS, ROLE_ADMIN, CREATE USER ON *.* TO 'teleport-admin' ;

CREATE DATABASE IF NOT EXISTS `teleport`;
GRANT ALTER ROUTINE, CREATE ROUTINE, EXECUTE ON `teleport`.* TO 'teleport-admin' ;
```
</TabItem>
</Tabs>

Users created by Teleport will be assigned the `teleport-auto-user` role in the
database, which will be created automatically if it doesn't exist.

Teleport will not delete the automatically created user at the end of the
session. Instead, the user will be stripped of all roles and the user account
will be locked.

> Next, enable the database admin on the Teleport database configuration:

Reviewer: ```suggestion
Next, enable the database admin on the Teleport Database Service configuration:
```

<Tabs>
<TabItem label="Static config">
```yaml
db_service:
  enabled: "yes"
  databases:
  - name: "example"
    protocol: "mysql"
    uri: "localhost:3306"
    admin_user:
      name: "teleport-admin"
```
</TabItem>
<TabItem label="Dynamic resource">
```yaml
kind: db
version: v3
metadata:
  name: example
spec:
  protocol: "mysql"
  uri: "localhost:3306"
  admin_user:
    name: "teleport-admin"
```
</TabItem>
</Tabs>

<Admonition type="tip" title="Auto-discovered databases">
For auto-discovered cloud databases, the name of the admin user is taken from
the `teleport.dev/db-admin` label.
</Admonition>

## Step 2/3. Configure Teleport role

(!docs/pages/includes/database-access/auto-user-provisioning-common-teleport-role.mdx!)

> User created within the database will:

Reviewer: ```suggestion
Users created within the database will:
```

- Be assigned the `teleport-auto-user` role.
- Be assigned all roles from the Teleport user's role set that match the database.
  The role names must be valid and exist in the database.

> MySQL limits username up to 32 characters long. When the Teleport username is

Reviewer: ```suggestion
MySQL limits usernames to 32 characters. When the Teleport username is
```

within this limit, the user created within the database will have the same name

> as the Teleport username.
> 
> When the Teleport username is over the 32 characters limit, the user created

Reviewer: ```suggestion
as the Teleport username. When the Teleport username is over the 32 character limit, the user created
```
